File: client/client/scripts/dorado/widget/form/custom-dropdown.js
/**
 * @author Benny Bao (mailto:benny.bao@bstek.com)
 * @component Trigger
 * @class 自定义下拉框。
 * @extends dorado.widget.DropDown
 */
dorado.widget.CustomDropDown = $extend(dorado.widget.DropDown,/** @scope dorado.widget.CustomDropDown.prototype */ {
	$className: "dorado.widget.CustomDropDown",
	
	ATTRIBUTES: /** @scope dorado.widget.CustomDropDown.prototype */ {
	
		/**
		 * 要显示在自定义下拉框中的控件。
		 * @type dorado.widget.Control
		 * @attribute writeBeforeReady
		 */
		control: {
			writeBeforeReady: true,
			innerComponent: ""
		}
	},
	
	createDropDownBox: function(editor, dropDown) {
		var box = $invokeSuper.call(this, arguments);
		var control = this._control;
		box.set("control", control);
		box.addListener("beforeShow", function() {			
			var $box = jQuery(box.getDom().firstChild), boxWidth = $box.width(), boxHeight = $box.height();
			var $dom = jQuery(control.getDom()), realWidth = $dom.outerWidth(), realHeight = $dom.outerHeight(), shouldRefresh;
			if (realWidth < boxWidth) {
				control.set("width", boxWidth);
				shouldRefresh = true;
			}
			if (realHeight < boxHeight) {
				control.set("height", boxHeight);
				shouldRefresh = true;
			}
			if (shouldRefresh) control.refresh();
		});
		return box;
	}
});

File: view/src/com/bstek/dorado/view/widget/form/trigger/CustomDropDown.java
package com.bstek.dorado.view.widget.form.trigger;

import com.bstek.dorado.annotation.ClientObject;
import com.bstek.dorado.annotation.ClientProperty;
import com.bstek.dorado.annotation.XmlSubNode;
import com.bstek.dorado.view.annotation.Widget;
import com.bstek.dorado.view.widget.Control;
import com.bstek.dorado.view.widget.InnerElementReference;

/**
 * @author Benny Bao (mailto:benny.bao@bstek.com)
 * @since 2010-8-10
 */
@Widget(name = "CustomDropDown", category = "Trigger",
		dependsPackage = "base-widget")
@ClientObject(prototype = "dorado.widget.CustomDropDown",
		shortTypeName = "CustomDropDown")
public class CustomDropDown extends DropDown {
	private InnerElementReference<Control> controlRef = new InnerElementReference<Control>(
			this);

	@XmlSubNode
	@ClientProperty
	public Control getControl() {
		return controlRef.get();
	}

	public void setControl(Control control) {
		controlRef.set(control);
	}
}
